### PR TITLE
feat(shouldDropdownOpen): introduce option to open the menu programmatically

### PR DESCRIPTION
**Pull request by s-pace**
- feat(dropdown): let user override shouldDropdownOpen based on the Autocomplete's state
- story(classNames): remove superfluous debugging log from #5
- refactor(Autocomplete): revert wrong new statement
- story(shouldDropdownOpen): add dedicated story
- refactor(gitignore): ignore IDEA folders
- doc(fix): add missing doc from #5

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -363,6 +363,12 @@ Called when the input changes.
 
 This turns experience is "controlled" mode. You'll be in charge of updating the state with the [top-level API](#autocomplete).
 
+#### `shouldDropdownOpen`
+
+> `(options: { state: AutocompleteState }) => boolean`
+
+Needed to check if the dropdown should open depending on the Autocomplete state.
```
Comment (francoischalifour) [inline README.md]:
```suggestion
Called to check whether the dropdown should open based on the Autocomplete state.

The default behavior is to open the dropdown when there are results.
```

File: README.md [hunk 2]
```diff
@@ -517,6 +523,12 @@ const source = {
 
 Called when an item is selected.
 
+#### `classNames`
+
+> `ClassNames`
+
+CSS classes applied to the template after the source.
```
Comment (francoischalifour) [inline README.md]:
```suggestion
CSS classes applied to the template of the source.
```

File: README.md [hunk 3]
```diff
@@ -363,6 +363,12 @@ Called when the input changes.
 
 This turns experience is "controlled" mode. You'll be in charge of updating the state with the [top-level API](#autocomplete).
 
+#### `shouldDropdownOpen`
+
+> `(options: { state: AutocompleteState }) => boolean`
```
Comment (francoischalifour) [inline README.md]:
```suggestion
> `(options: { state: AutocompleteState }) => boolean` | defaults to `({ state }) => state.results.some(result => result.suggestions.length > 0)`
```

File: src/types.ts [hunk 4]
```diff
@@ -318,6 +318,10 @@ export interface PublicAutocompleteProps {
       query: string;
     }
   ) => void | Promise<void | { state: AutocompleteState }>;
+  /**
+   * Called to know if the dropdown should be opened depending on the state.
```
Comment (francoischalifour) [inline src/types.ts]:
```suggestion
   * Called to check whether the dropdown should open based on the Autocomplete state.
```

File: stories/options.stories.tsx [hunk 5]
```diff
@@ -1207,6 +1202,248 @@ storiesOf('Options', module)
         },
       });
 
+      return container;
+    })
+  )
+  .add(
+    'shouldDropdownOpen=()=>true and minLength=4',
```
Comment (francoischalifour) [inline stories/options.stories.tsx]: We can simplify this story so there is less boilerplate to demo the `shouldDropdownOpen` feature.

**Review by francoischalifour — APPROVED** [review]

File: stories/options.stories.tsx (hunk 5 above)
Comment (s-pace) [reply stories/options.stories.tsx]: How would you like to simplify it? It seems to be already quite simple and visual. Let me know what are your thoughts on it.

Comment (francoischalifour) [reply stories/options.stories.tsx]: I mean code-wise, it's very verbose. We could remove multiple sources and SVGs to demo the feature without overhead. The only interesting line is:

```ts
shouldDropdownOpen: () => true,
```

Comment (francoischalifour) [reply stories/options.stories.tsx]: I updated the story in 3e7de6f7e38c70ec4ee40de0d81ffca282d98593.

**Comment by francoischalifour** [comment]
We'll need to find a story that _doesn't_ open the menu when it first loads the results so we don't have an empty dropdown open.

The following could work but it flashes when there are no results and you keep typing.

```ts
autocomplete({
  //...
  shouldDropdownOpen({ state }) {
    return (
      !state.isLoading ||
      state.results.some(result => result.suggestions.length > 0)
    );
  },
});
```

Let's tackle that in another PR.

**Pull request merged**